src/sentry/snuba/metrics/utils.py

```diff
@@ -324,15 +327,78 @@ class OrderByNotSupportedOverCompositeEntityException(NotSupportedOverCompositeE
     ...
 
 
-def get_intervals(start: datetime, end: datetime, granularity: int, interval: Optional[int] = None):
+def to_intervals(
+    start: datetime, end: datetime, interval_seconds: int
+) -> Tuple[datetime, datetime, int]:
+    """
+    Given a start date, end date and an alignment interval in seconds returns the aligned start, end and
+    the number of total intervals in [start:end]
+
+    """
+    assert interval_seconds > 0
+
+    if start.tzinfo is None:
+        start.replace(tzinfo=timezone.utc)
+    if end.tzinfo is None:
+        end.replace(tzinfo=timezone.utc)
+
+    interval_start = int(start.timestamp() / interval_seconds) * interval_seconds
+    interval_end = end.timestamp()
+
+    seconds_to_cover = interval_end - interval_start
+
+    last_incomplete_interval = 0
+    # NOTE: currently a [start,end] == [10:30, 11:30] generates [10:00,11:00]
+    # if instead  we want to generate [10:00,11:00],[11:00,12:00] then uncomment if below.
+    # if seconds_to_cover % interval_seconds != 0:
+    #     # we don't finish neatly at the end of interval, add another
+    #     # interval to cover the last incomplete period
+    #     last_incomplete_interval = 1
+
+    num_intervals = int(seconds_to_cover / interval_seconds) + last_incomplete_interval
+    # finally convert back to dates
+    adjusted_start = datetime.fromtimestamp(interval_start, timezone.utc)
+    adjusted_end = adjusted_start + timedelta(seconds=interval_seconds * num_intervals)
+    return adjusted_start, adjusted_end, num_intervals
+
+
+def get_num_intervals(
+    start: datetime,
+    end: datetime,
+    granularity: int,
+    interval: Optional[int] = None,
+) -> int:
+    """
+    Calculates the number of intervals from start to end.
+    If start==None then it calculates from the beginning of unix time (for backward compatibility with
+    MetricsQuery.calculate_intervals_len)
+    """
+
+    # legacy usage (if no start time assume beginning of Unix time)
+    if start is None:
+        start = datetime.fromtimestamp(0).replace(tzinfo=timezone.utc)
+
     if interval is None:
-        assert granularity > 0
-        delta = timedelta(seconds=granularity)
-    else:
-        start = datetime.fromtimestamp(int(start.timestamp() / interval) * interval, timezone.utc)
-        end = datetime.fromtimestamp(int(end.timestamp() / interval) * interval, timezone.utc)
-        assert interval > 0
-        delta = timedelta(seconds=interval)
-    while start < end:
+        interval = granularity
+
+    assert interval > 0
+
+    _start, _end, num_intervals = to_intervals(start, end, interval)
+    return num_intervals
+
+
+def get_intervals(
+    start: datetime, end: datetime, granularity: int, interval: Optional[int] = None
+) -> Generator[datetime, None, None]:
+    if interval is None:
+        interval = granularity
+
+    start, end, num_intervals = to_intervals(start, end, interval)
```

Comment:
If `end` is not used, I would explicitly show it.

```suggestion
    start, _end, num_intervals = to_intervals(start, end, interval)
```

Reply:
Yep, thanks !